packages/server/src/main/java/software/uncharted/terarium/hmiserver/utils/rebac/ReBACService.java

```diff
@@ -77,12 +84,51 @@ public class ReBACService {
 
 	public static String API_SERVICE_USER_ID;
 
-	volatile String CURRENT_ZED_TOKEN;
+	private volatile String CURRENT_ZED_TOKEN;
 
 	private String getKeycloakBearerToken() {
 		return "Bearer " + keycloak.tokenManager().getAccessTokenString();
 	}
 
+	private class CacheKey {
+
+		SchemaObject who;
+		Schema.Permission permission;
+		SchemaObject what;
+
+		CacheKey(SchemaObject who, Schema.Permission permission, SchemaObject what) {
+			this.who = who;
+			this.permission = permission;
+			this.what = what;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if (!(o instanceof CacheKey)) {
+				return false;
+			}
+			CacheKey other = (CacheKey) o;
+			return who.equals(other.who) && permission == other.permission && what.equals(other.what);
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(who, permission, what);
+		}
+	}
+
+	private final Cache<CacheKey, Boolean> permissionCache = Caffeine.newBuilder()
+		.expireAfterWrite(5, TimeUnit.MINUTES)
+		.recordStats()
+		.removalListener((Object key, Object value, RemovalCause cause) -> log.info("Key {} was removed {}", key, cause))
```

Comment:
Should this be a the `debug` level and not `info`?
Would thise be everywhere in the logs?